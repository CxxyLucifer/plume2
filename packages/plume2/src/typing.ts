import { Map } from 'immutable';
import { ViewAction } from './view-action';

/**
 * Immutable的map类型
 */
export type IMap = Map<string, any>;

/**
 * actor的路由类型
 */
export type TRoute = {
  [name: string]: (state: IMap, params?: any) => IMap;
};

/**
 * Relax容器组件的上下文类型
 */
export interface IRelaxContext<Store> {
  _plume$Store: Store;
}

/**
 * Relax容器组件的props类型
 */
export interface IRelaxProps {
  relaxProps?: Object;
  [name: string]: any;
}

/**
 * Relax容器组件的类型
 */
export interface IRelaxComponent extends React.ComponentClass<IRelaxProps> {
  relaxProps?: Object;
}

/**
 * Store数据容器的参数配置类型
 */
export interface IOptions {
  debug?: boolean;
  [name: string]: any;
}

export interface IViewActionMapper {
  [name: string]: ViewAction;
}

export type TViewAction<T> = {
  [K in keyof T]: T[K] extends new (...args: Array<any>) => infer R ? R : any
};
