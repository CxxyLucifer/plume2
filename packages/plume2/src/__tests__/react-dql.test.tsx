import * as React from 'react';
import * as renderer from 'react-test-renderer';
import { Map } from 'immutable';
import { Actor, Store, StoreProvider, QL, Relax } from '../index';
import { IMap } from '../typing';

class ProductActor extends Actor {
  defaultState() {
    return {
      products: [
        { id: 1, name: 'p1' },
        { id: 2, name: 'p2' },
        { id: 3, name: 'p3' },
        { id: 4, name: 'p4' },
        { id: 5, name: 'p5' }
      ]
    };
  }
}

class AppStore extends Store {
  bindActor() {
    return [new ProductActor()];
  }
}

@StoreProvider(AppStore)
class ProductApp extends React.Component {
  store: Store;

  render() {
    const products = this.store.state().get('products');

    return (
      <div>
        {products.map((p, index) => {
          return <ProductItem index={index} key={p.get('id')} />;
        })}
      </div>
    );
  }
}

const productQL = QL('productQL', [
  ['products', '$index'],
  /**
   * 
   */
  p => p
]);

@Relax
class ProductItem extends React.Component {
  static defaultProps = {
    index: 0
  };

  static relaxProps = {
    product: productQL
  };

  props: {
    index: boolean;
    relaxProps?: {
      product: IMap;
    };
  };

  render() {
    const { id, name } = this.props.relaxProps.product.toJS();
    return (
      <div>
        <div>{id}</div>
        <div>{name}</div>
      </div>
    );
  }
}

describe('relax dql', () => {
  it('initialize', () => {
    const tree = renderer.create(<ProductApp />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
