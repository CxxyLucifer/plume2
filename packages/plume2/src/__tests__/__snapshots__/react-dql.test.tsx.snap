// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`relax dql initialize 1`] = `
<div>
  <div>
    <div>
      1
    </div>
    <div>
      p1
    </div>
  </div>
  <div>
    <div>
      2
    </div>
    <div>
      p2
    </div>
  </div>
  <div>
    <div>
      3
    </div>
    <div>
      p3
    </div>
  </div>
  <div>
    <div>
      4
    </div>
    <div>
      p4
    </div>
  </div>
  <div>
    <div>
      5
    </div>
    <div>
      p5
    </div>
  </div>
</div>
`;
